corehq/apps/sso/models.py

```diff
@@ -271,6 +284,20 @@ def __str__(self):
         return f"{self.email_domain} authenticated by [{self.identity_provider.name}]"
 
 
+@receiver(pre_save, sender=AuthenticatedEmailDomain)
+@receiver(pre_delete, sender=AuthenticatedEmailDomain)
+def clear_caches_for_email_domain(sender, instance, **kwargs):
```

Comment:
> Delete signals work on the delete queryset method. If you'd like the same actions to happen on both single object and queryset deletes, a signal is probably your best bet. 

this is certainly one valid use of why it's needed. Additionally, this doesn't modify the model itself (as would be assumed with the `delete` and `save` methods, but is related to another model)

Reply:
Thanks for pointing that out. I'm now on the fence. I still think the arguments in that article hold -- signals make the code harder to follow. Regardless if a model needs to update itself or a different model, I'd always prefer seeing the save/delete code in the expected save/delete methods, rather than having to know that a signal exists somewhere. The queryset stuff is a real pain, though. The one spot where we're doing a `filter()` into a `delete()` could be handled via a `get()` and then a `delete()`, but filter is such a common pattern that I think relying on developers to always use `get` is about as likely to cause errors and confusion as signals are. I'd like to hear others' thoughts on this.  If we do opt to stay with signals, is there anything we can add to follow that article's closing thoughts of `By documenting them copiously, using some logging, and leave lots of breadcrumbs you can use them safely and prevent future frustration.`?

Reply:
I think the documentation + placement of the signals next to the models make this fairly easy to follow. The models themselves are very short, so I'm pretty confident this wouldn't be missed. The clear cache on IdentityProvider's save I agree now should be in the `save` instance method, but I still don't see the argument on `AuthenticatedEmailDomain` and `TrustedIdentityProvider`, especially since I'm now going on the `post_save` and `post_delete` signals per Daniel's suggestion.

Reply:
hmm i think i'm also forgetting another cross-app signal for `Subscription` which would mean now it makes sense for signals to live in its own file vs models. 🤔

Reply:
ignore my previous comment about the separate signals file...still not going down that route. see latest commits